### PR TITLE
Audit updates

### PR DESCRIPTION
**Pull request by frett**
@hlbraddock

### CODE REVIEW

**Comment by frett** [comment]
I wanted a bit more information in the Audit messages for updating a user

File: api/src/main/java/org/ccci/idm/user/Group.java [hunk 1]
```diff
@@ -52,8 +53,10 @@ public int hashCode() {
         return Arrays.deepHashCode(new Object[]{this.path, this.name});
     }
 
+    private static final Joiner PATH_JOINER = Joiner.on(".").skipNulls();
+
     @Override
     public String toString() {
-        return MoreObjects.toStringHelper(this).add("path", this.path).add("name", this.name).toString();
+        return MoreObjects.toStringHelper(this).add("path", PATH_JOINER.join(path)).add("name", name).toString();
```
Comment (frett) [inline api/src/main/java/org/ccci/idm/user/Group.java]: before this change groups would stringify as `Group{path=[Ljava.lang.String;@5212494d, name=userAdmins}` now the path is actually included.

Comment (hlbraddock) [inline api/src/main/java/org/ccci/idm/user/Group.java]: +1

File: inspektr/src/main/java/org/ccci/idm/user/inspektr/spi/UpdateUserAuditResourceResolver.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
+package org.ccci.idm.user.inspektr.spi;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.MoreObjects.ToStringHelper;
+import org.ccci.idm.user.User;
+import org.ccci.idm.user.User.Attr;
+
+import javax.annotation.Nonnull;
+
+public class UpdateUserAuditResourceResolver extends UserAuditResourceResolver {
```
Comment (hlbraddock) [inline inspektr/src/main/java/org/ccci/idm/user/inspektr/spi/UpdateUserAuditResourceResolver.java]: Where is this class used?

Comment (frett) [inline inspektr/src/main/java/org/ccci/idm/user/inspektr/spi/UpdateUserAuditResourceResolver.java]: https://github.com/GlobalTechnology/thekey/commit/55f48f3c0fb00c12123731214dd5665ea519b2d2#diff-2bfc497d36504a12638c16931ffc393dR29

**Comment by hlbraddock** [comment]
This looks fine. +1

I don't quite know how a `User` object is passed to the `createResource(...)` method, for example, but I presume the configuration in https://github.com/GlobalTechnology/thekey/commit/55f48f3c0fb00c12123731214dd5665ea519b2d2#diff-2bfc497d36504a12638c16931ffc393dR29 is used by the inspektr audit framework in some `User` aware context.

**Comment by frett** [comment]
how these classes are used by inspektr is through an Aspect Oriented Programming pattern.

There is an `AuditTrailManagementAspect` class that spring automatically wires around any objects/methods that use the `@Audit` annotation. Whenever those methods are called `AuditTrailManagementAspect.handleAuditTrail()` is triggered immediately after the method returns (before the calling code even gets the response) and has access to all arguments passed to the method, as well as the response or exception thrown. It is within that method that it performs an implementation lookup for instances of `AuditResourceResolver` and triggers the `resolveFrom` method, which in turn triggers `createResource` in our implementations.

**Pull request merged**